### ISSUE TITLE
npm dependencies issue with pre-commit 2.8.1 ?

### ISSUE THREAD

**Issue by sbidoul** [pre-commit/pre-commit#1674]
Hi,

Since 2.8.1 one of our node-based pre-commit hook started to fail. With 2.7.1 it works fine.

To reproduce, create a git repo with an empty `test.js` file and the following `.pre-commit-config.yaml`:

```yaml
default_language_version:
  node: "14.15.0"
repos:
  - repo: https://github.com/prettier/prettier
    rev: 2.1.2
    hooks:
      - id: prettier
        name: prettier + plugin-xml
        additional_dependencies:
          - "@prettier/plugin-xml@0.12.0"
        args:
          - --plugin=@prettier/plugin-xml
```

`pre-commit run -a` then fails with `Error: Cannot find module 'prettier/doc'`. Cleaning the pre-commit cache and running the same with pre-commit 2.7.1 works fine.

```
[INFO] Initializing environment for https://github.com/prettier/prettier.
[INFO] Initializing environment for https://github.com/prettier/prettier:@prettier/plugin-xml@0.12.0.
[INFO] Installing environment for https://github.com/prettier/prettier.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
prettier + plugin-xml....................................................Failed
- hook id: prettier
- exit code: 1

internal/modules/cjs/loader.js:883
  throw err;
  ^

Error: Cannot find module 'prettier/doc'
Require stack:
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/@prettier/plugin-xml/src/embed.js
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/@prettier/plugin-xml/src/plugin.js
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/common/load-plugins.js
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/index.js
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/cli/index.js
- /home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/bin/prettier.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/@prettier/plugin-xml/src/embed.js:1:29)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/@prettier/plugin-xml/src/embed.js',
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/@prettier/plugin-xml/src/plugin.js',
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/common/load-plugins.js',
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/index.js',
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/src/cli/index.js',
    '/home/me/.cache/pre-commit/repo_znxgdpn/node_env-14.15.0/lib/node_modules/prettier/bin/prettier.js'
  ]
}
```

**Comment by MarcoGorelli** [comment]
Hi @sbidoul  - have you tried using prettier/pre-commit?

**Comment by sbidoul** [comment]
@MarcoGorelli I did not know about it. I just tried now and can't seem to make it work with the prettier-xml plugin.

**Comment by asottile** [comment]
I can't reproduce, what platform are you on?

```console
$ pre-commit run --all-files
[INFO] Initializing environment for https://github.com/prettier/prettier.
[INFO] Initializing environment for https://github.com/prettier/prettier:@prettier/plugin-xml@0.12.0.
[INFO] Installing environment for https://github.com/prettier/prettier.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
prettier + plugin-xml....................................................Failed
- hook id: prettier
- files were modified by this hook

t.js

$ pre-commit run --all-files
prettier + plugin-xml....................................................Passed
$ cat .pre-commit-config.yaml 
default_language_version:
  node: "14.15.0"
repos:
  - repo: https://github.com/prettier/prettier
    rev: 2.1.2
    hooks:
      - id: prettier
        name: prettier + plugin-xml
        additional_dependencies:
          - "@prettier/plugin-xml@0.12.0"
        args:
          - --plugin=@prettier/plugin-xml
```

**Comment by sbidoul** [comment]
I reproduce ubuntu 20 with pre-commit 2.8.1 installed with pipx. We discovered the issue on travis builds running on xenial (example: https://travis-ci.com/github/OCA/l10n-romania/jobs/421130459).

**Comment by asottile** [comment]
oh ok I can reproduce now, very strange 🤔

**Comment by asottile** [comment]
this is probably #1667 if I had to guess

**Comment by asottile** [comment]
ok I've figured this out -- it looks like this was only working by accident before :(

prettier has prettier as a dev dependency and so the old way was installing all of the dev dependencies and pulling in prettier's built distribution (unrelated to the actual version being installed!)

this is no longer happening and now the install is more like `npm install git+...`

it turns out that prettier doesn't install correctly via `npm install git+...` so their hook is just broken

fortunately, and they're in the progress of migrating this anyway, the `prettier/pre-commit` repository doesn't have this limitation

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 0501dd9..18f645f 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,12 +1,13 @@
 default_language_version:
   node: "14.15.0"
 repos:
-  - repo: https://github.com/prettier/prettier
-    rev: 2.1.2
+  - repo: https://github.com/prettier/pre-commit
+    rev: v2.1.2
     hooks:
       - id: prettier
         name: prettier + plugin-xml
         additional_dependencies:
+          - "prettier@2.1.2"
           - "@prettier/plugin-xml@0.12.0"
         args:
           - --plugin=@prettier/plugin-xml
```

so tl;dr is switch to prettier/pre-commit as their repository isn't correctly `git` installable

**Comment by sbidoul** [comment]
@asottile Thanks for the analysis and detailed explanation.
